File: framework/languages/cpp/e2e/RegistrationCodec/Client/main.cpp
/* SPDX-License-Identifier: FSL-1.1-ALv2 */

#include "Scenarios/auto_registration_scenario.hpp"
#include "Scenarios/attribute_registration_scenario.hpp"
#include "Scenarios/codec_mismatch_scenario.hpp"
#include "Scenarios/manual_registration_scenario.hpp"
#include "Scenarios/rc_a4_di_lifecycle_scenario.hpp"
#include "Scenarios/rc_a5_filter_ordering_scenario.hpp"
#include "Scenarios/rc_a6_invalid_registration_scenario.hpp"
#include "Scenarios/rc_b1_json_codec_scenario.hpp"
#include "Scenarios/rc_b2_protobuf_codec_scenario.hpp"
#include "Scenarios/rc_b3_messagepack_codec_scenario.hpp"
#include "Scenarios/rc_b4_codec_coexistence_scenario.hpp"

#include <iostream>
#include <stdexcept>
#include <string>

namespace rc = zlink::framework::e2e::registration_codec;
namespace rc_client = zlink::framework::e2e::registration_codec::client;

int main (int argc, char **argv)
{
    try {
        const auto options = rc_client::read_client_options (argc, argv);
        const auto &scenario = options.scenario;
        if (scenario == "b5" || scenario == "rc-b5") {
            rc_client::run_codec_mismatch_scenario (options);
        } else {
            bool ran = false;
            auto wants = [&scenario] (const char *id) {
                return scenario == "all" || scenario == id;
            };
            if (wants ("rc-a1")) {
                ran = true;
                rc_client::run_auto_registration_scenario (options);
            }
            if (wants ("rc-a2")) {
                ran = true;
                rc_client::run_attribute_registration_scenario (options);
            }
            if (wants ("rc-a3")) {
                ran = true;
                rc_client::run_manual_registration_scenario (options);
            }
            if (wants ("rc-a4")) {
                ran = true;
                rc_client::run_di_lifecycle_scenario (options);
            }
            if (wants ("rc-a5")) {
                ran = true;
                rc_client::run_filter_ordering_scenario (options);
            }
            if (wants ("rc-a6")) {
                ran = true;
                rc_client::run_invalid_registration_scenario (options);
            }
            if (wants ("rc-b1") || wants ("rc-b2") || wants ("rc-b3")) {
                const auto roundtrip = rc_client::post_empty<rc::codec_roundtrip_scenario_res_t> (
                  options.http_endpoint, "/codec/roundtrip");
                if (wants ("rc-b1")) {
                    ran = true;
                    rc_client::run_json_codec_scenario (options, roundtrip);
                }
                if (wants ("rc-b2")) {
                    ran = true;
                    rc_client::run_protobuf_codec_scenario (options, roundtrip);
                }
                if (wants ("rc-b3")) {
                    ran = true;
                    rc_client::run_messagepack_codec_scenario (options, roundtrip);
                }
            }
            if (wants ("rc-b4")) {
                ran = true;
                rc_client::run_codec_coexistence_scenario (options);
            }
            if (!ran) {
                throw std::runtime_error ("unknown RegistrationCodec scenario: " + scenario);
            }
        }
    }
    catch (const std::exception &error) {
        std::cerr << "registration-codec scenario failed: " << error.what () << "\n";
        return 1;
    }
    return 0;
}

File: framework/languages/node/e2e/RegistrationCodec/Client/main.ts
import { parseClientOptions } from './Support/client-options';
import { runRcA2 } from './Scenarios/AttributeRegistrationScenario';
import { runRcA1 } from './Scenarios/AutoRegistrationScenario';
import { runRcB5 } from './Scenarios/CodecMismatchScenario';
import { runRcA6 } from './Scenarios/InvalidRegistrationScenario';
import { runRcA3 } from './Scenarios/ManualRegistrationScenario';
import { runRcA4 } from './Scenarios/RcA4DiLifecycleScenario';
import { runRcA5 } from './Scenarios/RcA5FilterOrderingScenario';
import { runRcB1 } from './Scenarios/RcB1JsonCodecScenario';
import { runRcB2 } from './Scenarios/RcB2ProtobufCodecScenario';
import { runRcB3 } from './Scenarios/RcB3MessagePackCodecScenario';
import { runRcB4 } from './Scenarios/RcB4CodecCoexistenceScenario';
import { runRCB6 } from './Scenarios/rc-b6-scenario';

async function main(): Promise<void> {
  const options = parseClientOptions(process.argv.slice(2));
  const scenarios: Record<string, () => Promise<void>> = {
    'RC-A1': () => runRcA1(options.serverUrl),
    'RC-A2': () => runRcA2(options.serverUrl),
    'RC-A3': () => runRcA3(options.serverUrl),
    'RC-A4': () => runRcA4(options.serverUrl),
    'RC-A5': () => runRcA5(options.serverUrl),
    'RC-A6': () => runRcA6(options),
    'RC-B1': () => runRcB1(options.serverUrl),
    'RC-B2': () => runRcB2(options.serverUrl),
    'RC-B3': () => runRcB3(options.serverUrl),
    'RC-B4': () => runRcB4(options.serverUrl),
    'RC-B5': () => runRcB5(options.codecRequesterUrl, options.jsonOnlyUrl),
    'RC-B6': () => runRCB6(options.serverUrl)
  };
  if (options.scenario === 'all') {
    for (const run of Object.values(scenarios)) {
      await run();
    }
  } else {
    const run = scenarios[options.scenario];
    if (run === undefined) {
      throw new Error(`Unknown RegistrationCodec scenario '${options.scenario}'.`);
    }
    await run();
  }
  console.log('registration-codec e2e result=passed');
}

main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});
